src/js/structs/CompositeState.js

```diff
@@ -74,6 +74,23 @@ class CompositeState {
     enrichNodeDataWithHealthData(this.data.slaves, this.nodeHealthData);
   }
 
+  addMesosState(newData) {
+    if (this._isDisabled()) {
+      storedState = newData;
+
+      return;
+    }
+    if (newData == null) {
+      return;
+    }
+
+    this.data = { ...this.data, ...newData };
```

Comment:
The only difference with addState seems to be that this is preserving all keys. Which keys are missing and can they be added to the PRESERVE_KEYS array?

Reply:
i updated the commit and hopefully made things a lot more explicit. the convolution here was that someone wanted to be able to set all data via `addState`. we now handle `master_info` explicitly.